### DISCUSSION TITLE
Add child nodes to Interactive node views

### DISCUSSION THREAD

**Discussion by janus-reith** [ueberdosis/tiptap#2397] [General Questions]
I'm failing to create a custom component that has multiple editable areas.
I'm using React, but I assume that these considerations are not specific to the React Node views.

To have some visualization, let's assume this is the desired result:

![tiptap_mockup_A_B](https://user-images.githubusercontent.com/1831220/149743551-1d42e726-ba03-4402-83f5-fcd1a343ea40.png)

I know that we have the limitation of only one `NodeViewContent` per `NodeViewWrapper`. While that would most probably be the most simple implementation on my end if I could just place multiple `NodeViewContent's` per Wrapper with unique names, I'm okay to live with this limit and and make these editable areas separate components within, that each have one editable.
That leads me to the main question: Can I A. define other nodes to be set within my custom interactive node view at a specific position surrounded by whatever other static content I might have there , or else, can I B. define multiple areas where childNodes are rendered/in case of React, manually iterate over the `children` ?

I know that the following is possible:

Outer Node:
![tiptap_mockup_A_B_outer](https://user-images.githubusercontent.com/1831220/149744783-c9c8eb03-4da6-4c92-a83f-2ed4f37e9b03.png)
That component would be rather "dumb" at that point and contain nothing more than a simple wrapper

Inner Node A:
![tiptap_mockup_A_B_A](https://user-images.githubusercontent.com/1831220/149745325-0cf021b2-d2b3-4501-85eb-a863ffc67ded.png)
Only covers the one A column, and contains one editable text area.

Same for Inner Node B:
![tiptap_mockup_A_B_B](https://user-images.githubusercontent.com/1831220/149745608-cfd16f60-3a4a-45b6-9bf5-81c3036d48fc.png)
Only covers the one B column, and contains one editable text area.

I really want to avoid that kind of split-up since it would decrease the DX for my usecase a lot. Having the restriction of one editable content per wrapper is okay, but then I'd need a way to place these nodes within my node within my outer interactive node view. And if that also doesn't work, I guess a workaround might be to iterate over passed subnodes from my outer node, and have logic like: For child node 0, render it like this, passing my static content as a wrapper, for child node 1, do xyz.

Is that possible, or are there other options I might be missing? I appreciate any hints :)

**Comment by hanspagel** [comment]
You can’t add multiple editable areas to one Node, yes. So there’s no way to avoid splitting it up into multiple nodes.

> but then I'd need a way to place these nodes within my node within my outer interactive node view.

I don’t understand what you mean. That’s CSS, isn’t it?

**Reply by janus-reith** [reply]
Thanks for the response!

>>but then I'd need a way to place these nodes within my node within my outer interactive node view.

>I don’t understand what you mean. That’s CSS, isn’t it?

No what I mean is, within my React interactive view, I could add whatever content in there inside of the wrapper, but without it being relevant to the actual data in the tree that the editor manages.
Like above, e.g. "static content specific to A" - Which might be anything from simple text so some complex component tiptap knows nothing about and doesn't need to parse to/from html.

For a simple node with single editable content that works fine, e.g. I could easily create a node like "Inner Node A". However I don't think I would be able to create something like in the first picture, in a way that I can define "static content specific to A" and "static content specific to B" within one parent component.

**Reply by hanspagel** [reply]
Sorry, I still don’t get it. Can you share an example of how the HTML output would look like then?

You definitely can create a "parent node" that always starts with node A followed by node B. And node A and node B can have different static and editable content.

**Reply by janus-reith** [reply]
> You definitely can create a "parent node" that always starts with node A followed by node B. And node A and node B can have different static and editable content.

Yes, that would be possible - However, it would mean that I split up the two nodes A and B at the highest point in the DOM where they are each not nested in a different element.
And then for component A I can place static content in there and one editable element, and the same for B.

> Sorry, I still don’t get it. Can you share an example of how the HTML output would look like then?

Thanks for taking the time to try to understand it! :) The desired html might for example be this:
```
<div id="outerNode">
   <div id="ISadlyNeedToBeNodeA">
     <div className="IAmSomeStaticContent">Content</div>
     <>Editable Area</>
  </div>
   <div id="ISadlyNeedToBeNodeB">
     <div className="IAmSomeOtherStaticContent">Content</div>
     <>Editable Area</>
  </div>
</div>
```
(Although in my case, the `IAmSomeStaticContent` div's wouldn't even be relevant for the html output in my usecase, I added a different renderer for each component which returns MJML to create Emails and really just need the content of the editables).

A structure like this works:
Outer:
```
 <NodeViewWrapper>
  <div id="outerNode">
     <NodeViewContent />
  <div>
</NodeViewWrapper>
```

`NodeViewContent` is kind of a blackbox here, I can't nest anything here from the parent(which would also be a bit weird) - I can only define a `content: "NodeA NodeB"` field  and then e.g. make NodeA:

```
<NodeViewWrapper>
  <div id="ISadlyNeedToBeNodeA">
       <div className="IAmSomeStaticContent">Content</div>
        <NodeViewContent />
  </div>
</NodeViewWrapper>
```
and the same for NodeB.

What however does not work, is to have some kind of layout skeleton defined in my `outerNode`, which includes the "IAmSomeStaticContent" div's and two editable areas. (AFAIK, if that actually works that would solve my issue)

Now for that simple example above it doesn't seem like a huge issue to split up theses nodes, and in general I would also consider it good practice to keep the scope of your functions low. But in my case, with some components that needs to have multiple inputs with certain restrictions, that reaches a point where the DX get's quite bad, and I can't simply author certain sections in one go without ending up creating several unique components.

**Comment by janus-reith** [comment]
In case this won't work, I wonder if using several tiptap instances per page is an option, one per editable area?
So far I had ruled it out for performance reasons, but maybe there is less of an impact than I expect?

**Comment by dphuang2** [comment]
@janus-reith  What about using custom nodes + React portals?

**Reply by janus-reith** [reply]
Hm that could be interesting, could you elaborate further what you mean? Thanks!

**Reply by dphuang2** [reply]
> Hm that could be interesting, could you elaborate further what you mean? Thanks!

I gave up with this approach. I was thinking I could make nodes with editorcontent that accept a `containerRef` where I can use `ReactDOM.createPortal` to position them in the node that I want multiple contenteditable divs.

Instead I just went with embedding inner components that has multiple `EditorContent` components and storing their content as attributes in the parent node.
